app/src/main/java/com/hedvig/app/feature/home/ui/changeaddress/ChangeAddressActivity.kt

```diff
@@ -34,14 +52,118 @@ class ChangeAddressActivity : BaseActivity(R.layout.change_address_activity) {
-            }
+        observeViewModel()
+    }
+
+    private fun observeViewModel() {
+        viewModel.viewState.observe(this) { viewState ->
+            TransitionManager.beginDelayedTransition(binding.root)
+            setViewState(viewState)
         }
     }
 
+    private fun setViewState(viewState: ViewState): Any = when (viewState) {
+        Loading -> {
+            binding.spinner.loadingSpinner.show()
+            binding.contentScrollView.remove()
+        }
+        SelfChangeAddress -> setContent(
+            titleText = getString(R.string.moving_intro_title),
+            subtitleText = getString(R.string.moving_intro_description),
+            buttonText = getString(R.string.moving_intro_open_flow_button_text),
+            buttonIcon = null,
+            // TODO: Start Embark
+            onContinue = { }
+        )
+        ManualChangeAddress -> setContent(
+            titleText = getString(R.string.moving_intro_title),
+            subtitleText = getString(R.string.moving_intro_manual_handling_description),
+            buttonText = getString(R.string.moving_intro_manual_handling_button_text),
+            buttonIcon = R.drawable.ic_chat_white,
+            onContinue = { openChat() }
+        )
+        is ChangeAddressInProgress -> setUpcomingChangeContent(
+            titleText = getString(R.string.moving_intro_existing_move_title),
+            subtitleText = getString(R.string.moving_intro_existing_move_description),
+            buttonText = getString(R.string.moving_intro_manual_handling_button_text),
+            buttonIcon = R.drawable.ic_chat_white,
+            onContinue = { openChat() },
+            viewState.upcomingAgreementResult
+        )
+        is UpcomingAgreementError -> setContent(
+            titleText = getString(R.string.error_dialog_title),
+            subtitleText = when (viewState.error) {
+                NoContractsError -> "You do not have any contracts eligible for address change"
+                is GeneralError -> viewState.error.message ?: "Could not continue, please try again later"
+            },
+            buttonText = "Try again",
+            buttonIcon = null,
+            onContinue = { viewModel.reload() }
+        )
+        is SelfChangeError -> setContent(
+            titleText = getString(R.string.error_dialog_title),
+            subtitleText = viewState.error.message ?: "Could not continue, please try again later",
+            buttonText = "Try again",
+            buttonIcon = null,
+            onContinue = { viewModel.reload() }
+        )
+    }
+
+    private fun openChat() = startActivity(ChatActivity.newInstance(this, true))
+
+    private fun setContent(
+        titleText: String,
+        subtitleText: String,
+        buttonText: String?,
+        @DrawableRes buttonIcon: Int?,
+        onContinue: () -> Unit
+    ) = with(binding) {
+        spinner.loadingSpinner.remove()
+        contentScrollView.show()
+        image.show()
```

Comment:
```suggestion
        spinner.loadingSpinner.isVisible = false
        contentScrollView.isVisible = true
        image.isVisible = true
```

I'm partial to replacing the `show()`/`remove()`-extensions with the `isVisible`-extension, as the latter can lead to less code in certain cases, with the prime example being:

```kotlin
if (condition) {
    view.show()
} else {
    view.remove()
}

// vs

view.isVisible = condition
```

Admittedly, in this scenario the difference is small, but I think there are gains to be had by phasing out the former style with the latter entirely 😄

Reply:
I agree its better in the if/else example, but I still think 

`image.show()` reads better than `image.isVisible = true`. No harm in using both right?

Reply:
I'm on the fence - but we can keep it as is 🙂